### DISCUSSION TITLE
Rendering UIElements on WebAssembly via CanvasKit/Skia renderer

### DISCUSSION THREAD

**Discussion by kszicsillag** [unoplatform/uno#10209] [Ideas]
[Currently](https://platform.uno/docs/articles/uno-development/uno-internals-wasm.html)
> There is for the most part a 1:1 mapping between managed XAML elements and DOM elements. The element tag associated with a XAML type is set by a UIElement constructor overload which takes the tag as a string, with "div" as the default.

Are there any plans for an alternative renderer for wasm, that utilizes Skia/CanvasKit in order to achieve consistent and pixel perfect UI across platforms? Just like [Flutter](https://docs.flutter.dev/development/platform-integration/web/renderers).

**Comment by jeromelaban** [comment]
Working with Skia has its own set of challenging when dealing with platform integration. Rendering using a Canvas complexifies significantly the integration of Javascript based controls in the UI which a lot of our customers apps are using.

That being said, pixel perfectness is generally attainable with what Uno provides already with styling, like fluent or material design systems.

Nonetheless, we know that some types of apps can benefit from this type of rendering and the work we're currently doing on the Skia.Gtk backend points at that direction.

Is there a specific type of app you're thinking about?

**Reply by kszicsillag** [reply]
Thanks for the reply.
We did some experiments a while ago regarding recreating a simplified version of a Wikidata Web UI (like the [Reasonator](https://reasonator.toolforge.org/?q=Q42)) with Uno WASM. That is some crowded UI (by design) with large number of DOM elements. It was not very performant. We were wondering if the Canvas way of rendering would have some perf benefits vs. the DOM way.

**Reply by jeromelaban** [reply]
Thanks. It may have indeed performance improvements, yet performance is a continuous work in progress in all areas. We have many updates to integrate to get DOM interop faster. Did you try with .NET Wasm AOT as well?

**Reply by kszicsillag** [reply]
No, not yet, but we are intended to look into it. Thanks!
